### PR TITLE
Deprecate initialize memcache store with dalli client

### PR DESCRIPTION
**Pull request by aledustet**
Part of #47323
The full fix will be to remove support for the `Dalli::Client` option when initializing a `MemCacheStore`

### Motivation / Background

Following up on [#47323](https://github.com/rails/rails/issues/47323). Many options are not forwarded to the Dalli client when it is initialized from the `ActiveSupport::Cache::MemCacheStore`. This is to support a broader set of features powered by the implementation. When an instance of a client is passed on the initializer, it takes precedence, and we have no control over which attributes will be overridden or re-processed on the client side; this is by design and should remain as such to allow both projects to progress independently. 
Having this option introduces several potential bugs that are difficult to pinpoint and get multiplied by which version of the tool is used and how each evolves.
During the conversation on the issue, the `Dalli` client maintainer supports [deprecating](https://github.com/rails/rails/issues/47323#issuecomment-1424292456) this option.

### Detail

This Pull Request adds a deprecation warning any time the `ActiveSupport::Cache::MemCacheStore` is initialized with an instance of a `Dalli::Client` instead of providing the address(es) of the memcache server(s) and any additional options to let the `MemCacheStore` implementation handle the provisioning of the client.

Removing this implicit dependency will ensure each library can evolve separately and cements the usage of `Dalli::Client` as an [implementation detail](https://github.com/rails/rails/issues/21595#issuecomment-139815433)

We can not remove a supported feature overnight, so we should add a deprecation warning for the next minor release(7.2 at this time).

A constant on the `Cache` namespace was only used to restrict options passed to the `Dalli::Client` initializer that now lives on the `MemCacheStore` class.

### Additional information

The main issue with allowing this approach is regarding potential misbehaviors and bugs because of both implementations' different strategies when dealing with keys, namespaces, and compression.
One example of functionality has already diverged between the two and how [fixing it](https://github.com/aledustet/rails/pull/1/files) would further increase the coupling between both libraries.
A follow-up to this would still be an attempt to include the key truncation recovery mechanism on `ActiveSupport` to the `Dalli` client, but this will not be the only time both libraries stray, and it will be up to maintainers to catch up.

### Checklist

Before submitting the PR make sure the following are checked:

* [X] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [X] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [X] Tests are added or updated if you fix a bug or add a feature.
* [X] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: activesupport/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,9 @@
+*   Deprecated initializing a `ActiveSupport::Cache::MemCacheStore` with an instance of `Dalli::Client`.
+
+    Previously providing an instance of `Dalli::Client` to `ActiveSupport::Cache::MemCacheStore` was allowed, but the difference in behavior between `Dalli::Client` and `ActiveSupport::Cache::MemCacheStore` introduces some undocumented bugs. For example reading multiple keys, when the client instance has a custom namespace would not work as expected.
```
Comment (eileencodes) [inline activesupport/CHANGELOG.md:3]: I dont want to imply that if we document a bug it's ok 😆 

```suggestion
    Previously providing an instance of `Dalli::Client` to `ActiveSupport::Cache::MemCacheStore` was allowed, but the difference in behavior between `Dalli::Client` and `ActiveSupport::Cache::MemCacheStore` introduces inconsistent behavior. For example reading multiple keys, when the client instance has a custom namespace would not work as expected.
```

File: activesupport/lib/active_support/cache.rb [hunk 2]
```diff
@@ -18,10 +18,6 @@ module Cache
     autoload :NullStore,        "active_support/cache/null_store"
     autoload :RedisCacheStore,  "active_support/cache/redis_cache_store"
 
-    # These options mean something to all cache implementations. Individual cache
-    # implementations may support additional options.
-    UNIVERSAL_OPTIONS = [:namespace, :compress, :compress_threshold, :expires_in, :expire_in, :expired_in, :race_condition_ttl, :coder, :skip_nil]
```
Comment (eileencodes) [inline activesupport/lib/active_support/cache.rb:23]: We can't change the name and location of this constant without deprecation, it is in the public API https://edgeapi.rubyonrails.org/classes/ActiveSupport/Cache.html

We could move it and assign UNIVERSAL_OPTIONS to the new one, and then deprecate this one.

File: activesupport/lib/active_support/cache/mem_cache_store.rb [hunk 3]
```diff
@@ -108,6 +113,8 @@ def self.build_mem_cache(*addresses) # :nodoc:
       #
       # If no addresses are provided, but <tt>ENV['MEMCACHE_SERVERS']</tt> is defined, it will be used instead. Otherwise,
       # MemCacheStore will connect to localhost:11211 (the default memcached port).
+      # Passing a Dalli::Client instance instead of addresses, is deprecated
+      # and will be removed.
```
Comment (eileencodes) [inline activesupport/lib/active_support/cache/mem_cache_store.rb:116]:
```suggestion
      ```suggestion

      # Passing a +Dalli::Client+ instance is deprecated and will be removed. Please pass an address instead.
```
```

File: activesupport/lib/active_support/cache/mem_cache_store.rb [hunk 4]
```diff
@@ -120,12 +127,15 @@ def initialize(*addresses)
           raise ArgumentError, "First argument must be an empty array, an array of hosts or a Dalli::Client instance."
         end
         if addresses.first.is_a?(Dalli::Client)
+          ActiveSupport.deprecator.warn(<<~MSG)
+            Initializing MemCacheStore with a Dalli::Client is deprecated and will be removed in Rails 7.2.
```
Comment (eileencodes) [inline activesupport/lib/active_support/cache/mem_cache_store.rb:130]: If it's deprecated in favor of something else that should be mentioned in the deprecation notice. We can't guarantee someone will read the docs.

**Review by eileencodes — CHANGES_REQUESTED** [review]

**Comment by matthewd** [comment]
I don't quite understand how we get here from the original issue (beyond the fact it was suggested). 

> When an instance of a client is passed on the initializer, it takes precedence, and we have no control over which attributes will be overridden or re-processed on the client side; this is by design and should remain as such to allow both projects to progress independently.

"this is by design and should remain as such" -> but actually let's just remove it?

> Having this option introduces several potential bugs that are difficult to pinpoint and get multiplied by which version of the tool is used and how each evolves.

What are these several potential bugs, and how do they differ between versions?

> Removing this implicit dependency will ensure each library can evolve separately

We're still constructing instances of the same class, and assuming its API. Again, I don't follow what evolution this permits?

File: activesupport/lib/active_support/cache.rb (hunk 2 above)
Comment (aledustet) [reply activesupport/lib/active_support/cache.rb]: Thanks for the pointer 👍 I assume this would go on a separate PR?
I did a quick grep and couldn't see any examples of constants being deprecated. 
How could we wrap accessing the constant with a deprecation warning?

File: activesupport/lib/active_support/cache/mem_cache_store.rb (hunk 4 above)
Comment (aledustet) [reply activesupport/lib/active_support/cache/mem_cache_store.rb:129]: First time I hear of this phenomenon 😅

**Comment by aledustet** [comment]
> I don't quite understand how we get here from the original issue (beyond the fact it was suggested).

Part of my reasoning to go with the suggestion of deprecating, is that it is not documented explicitly, and given that the one fix I found(could definitely be other options) is a bit more involved than removing the possibility of initializing with a `Dalli::Client` instance. 
There wasn't much discourse other than the conversation on the issue so we for sure can be missing more than a few angles but I landed on this because it seemed like a fast way to stop the source of this issue and a couple more mentioned on the issue, like double compressing.

> "this is by design and should remain as such" -> but actually let's just remove it?

Sorry, I think my idea likely got cramped up in my attempt to edit it. 
I was trying to mention that the `non forwarding` of properties like, namespace, compress, and other configuration options supported by `Dalli::Client` are not forwarded by design. The reason being different strategies/behavior(like the bug on the issue) to key management, encoding and other base features from both approaches. Given that these features at are odds, depending on the configuration of the `Dalli::Client`, I suggest we remove the `option of injecting a fully initialized client`, instead of attempting to delegate more to it. The aspect that was by design is to delegate some portions but not all.

> What are these several potential bugs, and how do they differ between versions?

I can't point to a specific version. I was trying to make the point that if an internal, not documented behavior where to change, we could have more bugs than the one on the issue. This is always a possibility for external dependencies but having two ways we interact with the library depending on how it was initialized would be exacerbated.

> We're still constructing instances of the same class, and assuming its API. Again, I don't follow what evolution this permits?

I think that if we were to fix the bug in question, one potential solution would be to access the private API of the client, or expand the client's public API, increasing the contact surface for both libraries. This is what I am referring that will allow further evolution, because we would be removing one source of coupling. There is already one example of the wrong behavior when injecting the instance vs memcache_store initializing it's own. Removing the option of accepting a fully fleshed instance of the client will remove an entire set of behavior both(hopefully) libraries need to maintain.

File: activesupport/lib/active_support/cache.rb (hunk 2 above)
Comment (ghiculescu) [reply activesupport/lib/active_support/cache.rb]: I think you should do it in this PR. You could use [`deprecate_constant`](https://github.com/rails/rails/blob/f95c0b7e96eb36bc3efc0c5beffbb9e84ea664e4/activesupport/lib/active_support/deprecation/constant_accessor.rb#L5).

File: activesupport/CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,9 @@
+*   Deprecated initializing a `ActiveSupport::Cache::MemCacheStore` with an instance of `Dalli::Client`.
+
+    Previously providing an instance of `Dalli::Client` to `ActiveSupport::Cache::MemCacheStore` was allowed, but the difference in behavior between `Dalli::Client` and `ActiveSupport::Cache::MemCacheStore` introduces inconsistent behavior. For example reading multiple keys, when the client instance has a custom namespace would not work as expected.
```
Comment (jamiemccarthy) [inline activesupport/CHANGELOG.md:3]:
```suggestion
    Deprecate the undocumented option of providing an already-initialized instance of `Dalli::Client` to `ActiveSupport::Cache::MemCacheStore`. Such clients could be configured with unrecognized options, which could lead to unexpected behavior. Instead, provide addresses as documented.
```

I think it's clearer to use a more direct voice and conclude with an action item.

I tightened up the explanation in my second sentence, but I would also support removing that whole sentence. It's maybe too wordy for the ActiveSupport changelog. If people really care why it's being deprecated they can look up this PR :)

By the way, this option was added in Rails 3.0 (e2ed1a1ca, af44b07649), and then Rails 4.1 (98458eea42) changed it slightly while removing the documentation for it. So this has been an undocumented option since 2014.

File: activesupport/lib/active_support/cache/mem_cache_store.rb [hunk 6]
```diff
@@ -120,12 +125,16 @@ def initialize(*addresses)
...
+            Please pass an address instead.
```
Comment (jamiemccarthy) [inline activesupport/lib/active_support/cache/mem_cache_store.rb:130]:
```suggestion
            Use memcached server addresses instead.
```

This echoes the language in the first sentence of the docs.

File: activesupport/lib/active_support/cache/mem_cache_store.rb [hunk 7]
```diff
@@ -120,12 +125,16 @@ def initialize(*addresses)
           raise ArgumentError, "First argument must be an empty array, an array of hosts or a Dalli::Client instance."
```
Comment (jamiemccarthy) [inline activesupport/lib/active_support/cache/mem_cache_store.rb:125]:
```suggestion
          raise ArgumentError, "First argument must be an address or array of addresses."
```

1. Let's have this PR remove the suggestion of using a Dalli::Client instance.

and since we're doing that...

2. "An empty array" (or nil) is not going to be useful so let's not suggest that either. (I'm not sure why that's not an error, but that's out of scope.)
3. The docs call them "addresses" and not "hosts" so let's use that term.

File: activesupport/lib/active_support/cache.rb (hunk 2 above)
Comment (jamiemccarthy) [reply activesupport/lib/active_support/cache.rb]: Alejandro, I don't think this PR needs to do anything with this constant.

It's defined in this base module with the intention that all caches can have predictable options. It happens to currently only be used in the subclass `MemCacheStore` as part of its initialization, but that's an implementation detail. (I get that it's a little confusing, and if I were writing it from scratch, I would write it differently too.)

I don't think this constant should be deprecated, or renamed, or moved, and even if I'm wrong, I think this constant isn't related to the purpose of this PR. If we want to change something about the way any of the four cache subclasses pick their options for getting initialized, I think it's reasonable to open that in a separate issue/PR.

File: activesupport/lib/active_support/cache/mem_cache_store.rb (hunk 7 above)
Comment (jamiemccarthy) [reply activesupport/lib/active_support/cache/mem_cache_store.rb]: Oh I see, `nil` ends up being the same as `[]` due to a `.compact`:

https://github.com/rails/rails/blob/e184d33dfcef9e7265781ec84af559af4781765c/activesupport/lib/active_support/cache/mem_cache_store.rb#L93

Which ends up using Dalli's defaults:

https://github.com/petergoldstein/dalli/blob/8ad238df762b5bccf1b19343f712c8d1e23656cb/lib/dalli/servers_arg_normalizer.rb#L43

So "an empty array" (or nil) is fine. That was my misunderstanding, sorry.

File: activesupport/lib/active_support/cache/mem_cache_store.rb [hunk 8]
```diff
@@ -117,15 +122,19 @@ def initialize(*addresses)
         super(options)
 
         unless [String, Dalli::Client, NilClass].include?(addresses.first.class)
-          raise ArgumentError, "First argument must be an empty array, an array of hosts or a Dalli::Client instance."
+          raise ArgumentError, "First argument must be an address or array of addresses."
```
Comment (jamiemccarthy) [inline activesupport/lib/active_support/cache/mem_cache_store.rb:125]:
```suggestion
          raise ArgumentError, "First argument must be an empty array, address, or array of addresses."
```

Sorry, this was [my misunderstanding](https://github.com/rails/rails/pull/47340#discussion_r1102859342). Empty array is fine.

Comment (aledustet) [reply activesupport/lib/active_support/cache/mem_cache_store.rb:125]: :phew: I had started adding a test to not allow it and saw we can use the default hahaha, thanks for circling back 👍

**Comment by byroot** [comment]
For what it's worth, just last week I've seen an app doing:

```ruby
config.memcached_client = Dalli.new(...)
config.cache_store = MemCachedSore.new(
  config.memcached_client,
  pool_size: Integer(ENV["RAILS_MAX_THREADS"]),
  pool_timeout: 1,
  # ...
)
```

Not realizing that the pool arguments were noops. So at the very least I'm convinced we should warn when some parameters are passed but discarded.

**Comment by ghiculescu** [comment]
@aledustet could you please [squash your commits](https://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#squashing-commits)?

**Comment by aledustet** [comment]
All done @ghiculescu

**Comment by semaperepelitsa** [comment]
Is it recommended to initialize another instance of Dalli::Client if we need to use some methods not available through Rails.cache?
